### PR TITLE
[rand gen] split rand store and block queue to make share aggregation async

### PR DESCRIPTION
**Pull request by zekun000**

### CODE REVIEW

**Comment by zekun000** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11482?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `main`
  * **PR #11453** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11453?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #11464** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11464?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #11476** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11476?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #11482** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11482?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11482?utm_source=stack-comment).

File: consensus/src/rand/rand_gen/rand_manager.rs [hunk 1]
```diff
@@ -86,33 +94,35 @@ impl<S: Share, D: AugmentedData, Storage: AugDataStorage<D>> RandManager<S, D, S
...
-        );
-        self.rand_store
-            .lock()
-            .add_share(self_share)
+        let mut rand_store = self.rand_store.lock();
+        rand_store.add_rand_metadata(metadata.clone());
+        rand_store
+            .add_share(self_share.clone())
             .expect("Add self share should succeed");
+        self.network_sender
```
Comment (danielxiangzl) [inline consensus/src/rand/rand_gen/rand_manager.rs:124]: Why not broadcast first before try aggregation?

File: consensus/src/rand/rand_gen/rand_manager.rs [hunk 2]
```diff
@@ -86,33 +94,35 @@ impl<S: Share, D: AugmentedData, Storage: AugDataStorage<D>> RandManager<S, D, S
...
+            .broadcast_without_self(RandMessage::<S, D>::Share(self_share).into_network_message());
         self.spawn_aggregate_shares_task(metadata)
```
Comment (danielxiangzl) [inline consensus/src/rand/rand_gen/rand_manager.rs:126]: spawn_fetch_shares_task?

**Review by danielxiangzl — APPROVED** [review]

File: consensus/src/rand/rand_gen/rand_store.rs [hunk 3]
```diff
@@ -370,42 +340,47 @@ mod tests {
         }
     }
 
-    #[test]
-    fn test_rand_store() {
+    #[tokio::test]
+    async fn test_rand_store() {
```
Comment (zjma) [inline consensus/src/rand/rand_gen/rand_store.rs:344]: what requires async?

Comment (zekun000) [reply consensus/src/rand/rand_gen/rand_store.rs:344]: spawn_blocking needs to be in a runtime

File: consensus/src/rand/rand_gen/rand_store.rs [hunk 4]
```diff
@@ -42,18 +42,26 @@ impl<S: Share> ShareAggregator<S> {
...
-        Some(S::aggregate(
-            self.shares.values(),
-            rand_config,
-            rand_metadata,
-        ))
+        let rand_config = rand_config.clone();
+        let self_share = self
+            .get_self_share()
+            .expect("Aggregated item should have self share");
+        tokio::task::spawn_blocking(move || {
```
Comment (zjma) [inline consensus/src/rand/rand_gen/rand_store.rs:57]: [here](https://dtantsur.github.io/rust-openstack/tokio/sync/mpsc/struct.UnboundedSender.html#method.send) it says it never blocks... so no need of `spawn_blocking`?

Comment (zjma) [reply consensus/src/rand/rand_gen/rand_store.rs:57]: oh, i mean `decision_tx.send`

Comment (zjma) [reply consensus/src/rand/rand_gen/rand_store.rs:57]: nvm, it's for the aggregation

**Review by zjma — APPROVED** [review]

**Pull request merged**